packages/webpack-inline-constant-exports-plugin/index.js

```diff
@@ -0,0 +1,117 @@
+const HarmonyImportSideEffectDependency = require( 'webpack/lib/dependencies/HarmonyImportSideEffectDependency' );
+const HarmonyImportSpecifierDependency = require( 'webpack/lib/dependencies/HarmonyImportSpecifierDependency' );
+const ConstDependency = require( 'webpack/lib/dependencies/ConstDependency' );
+
+const moduleTypes = [ 'javascript/auto', 'javascript/dynamic', 'javascript/esm' ];
+
+function addConstantExport( module, name, value ) {
+	if ( ! module.constantExports ) {
+		module.constantExports = {};
+	}
+	module.constantExports[ name ] = value;
+}
+
+function removeDependency( module, dep ) {
+	module.removeDependency( dep );
+	dep.module.reasons = dep.module.reasons.filter( r => r.dependency !== dep );
+}
+
+class InlinePlugin {
+	constructor( matchers ) {
+		if ( Array.isArray( matchers ) ) {
+			this.matchers = matchers;
+		} else {
+			this.matchers = [ matchers ];
+		}
+	}
+
+	isConstantsModule( module ) {
+		return module && this.matchers.some( matcher => matcher.test( module.resource ) );
+	}
+
+	apply( compiler ) {
+		compiler.hooks.compilation.tap( 'InlinePlugin', ( compilation, { normalModuleFactory } ) => {
+			const handleParser = parser => {
+				const handleExport = ( statement, declaration ) => {
+					const { module } = parser.state;
+
+					if ( ! this.isConstantsModule( module ) ) {
+						return;
+					}
+
+					if (
+						statement.type === 'ExportNamedDeclaration' &&
+						declaration.type === 'VariableDeclaration' &&
+						declaration.kind === 'const'
+					) {
+						for ( const declarator of declaration.declarations ) {
+							if ( declarator.init.type === 'Literal' ) {
+								addConstantExport( module, declarator.id.name, declarator.init.raw );
+							}
+						}
+					}
+
+					if ( statement.type === 'ExportDefaultDeclaration' && declaration.type === 'Literal' ) {
+						addConstantExport( module, 'default', declaration.raw );
+					}
```

Comment:
This looks like it will handle all kinds of literals, do you know? Should we consider excluding array and object literals? I ask because it seems like this could introduce some unexpected behavior if we inline arrays and objects with the expectation that they're the same.

```js
export default []
//////
import arr from '⬆';
arr === arr; // false?
```